### PR TITLE
Add support for updating an object by primary key

### LINKED ISSUE #808: Update object by PrimaryKey
Something functionally equivalent to the Java [copyToRealmOrUpdate](https://realm.io/docs/java/1.2.0/api/io/realm/Realm.html#copyToRealmOrUpdate-E-) or Swift's [add(object, true)](https://realm.io/docs/swift/1.0.2/api/Classes/Realm.html#/s:FC10RealmSwift5Realm3addFTCS_6Object6updateSb_T_)

### PR DESCRIPTION
**Pull request by nirinchev**
TODO:
- [x] Handle adding to lists (`obj.Realm.Manage()` before `list.Add()`)
- [x] Add tests for lists
- [x] Add `[Ignore]`d tests that show that cycles are not handled as expected
- [x] Add XML docs that explain that we'll not resolve cycles
- [x] Tests for `TryGetPrimaryKey`

Resolves #808

### CODE REVIEW

File: Realm.Shared/Realm.cs [hunk 1]
```diff
@@ -492,6 +492,24 @@ internal SortDescriptorBuilder CreateSortDescriptorForTable(RealmObject.Metadata
...
+        /// <summary>
+        /// This realm will start managing a RealmObject which has been created as a standalone object.
+        /// </summary>
+        /// <param name="obj">Must be a standalone object, null not allowed.</param>
+        /// <param name="update">If true, and an object with the same primary key already exists, performs an update.</param>
+        /// <exception cref="RealmInvalidTransactionException">If you invoke this when there is no write Transaction active on the realm.</exception>
+        /// <exception cref="RealmObjectManagedByAnotherRealmException">You can't manage an object with more than one realm</exception>
+        public void Manage(RealmObject obj, bool update = false)
+        {
+            ManageInternal(obj, obj.GetType(), update);
```
Comment (fealebenpae) [inline Realm.Shared/Realm.cs]: If `obj` is null here there will be a `NullReferenceException` before `ManageInternal` throws `ArgumentNullException`.

**Comment by nirinchev** [comment]
@fealebenpae @kristiandupont feel free to take a look

File: Weaver/RealmWeaver.Fody/ModuleWeaver.cs [hunk 2]
```diff
@@ -965,18 +994,39 @@ private TypeDefinition WeaveRealmObjectHelper(TypeDefinition realmObjectType, Me
...
 
+                        managePlaceholder = il.Create(OpCodes.Nop);
+                        il.Append(managePlaceholder);
+
+                        il.Append(il.Create(OpCodes.Ldloc_0));
+                        il.Append(il.Create(OpCodes.Call, _realmObjectRealmGetter));
+                        il.Append(il.Create(OpCodes.Ldloc_0));
+                        il.Append(il.Create(OpCodes.Ldfld, field));
+                        il.Append(il.Create(OpCodes.Ldarg_2));
+                        il.Append(il.Create(OpCodes.Call, new GenericInstanceMethod(_realmManageGenericReference) { GenericArguments = { field.FieldType } }));
```
Comment (kristiandupont) [inline Weaver/RealmWeaver.Fody/ModuleWeaver.cs]: Couldn't this just as well be the non-generic method? Or is this to optimize performance?

File: Weaver/RealmWeaver.Fody/ModuleWeaver.cs [hunk 3]
```diff
@@ -965,18 +994,39 @@ private TypeDefinition WeaveRealmObjectHelper(TypeDefinition realmObjectType, Me
...
             {
-                var property = kvp.Key;
-                var field = kvp.Value;
+                var property = prop.Property;
+                var field = prop.Field;
 
                 if (property.SetMethod != null)
                 {
-                    if (!IsNullable(property))
+                    Instruction updatePlaceholder = null;
```
Comment (kristiandupont) [inline Weaver/RealmWeaver.Fody/ModuleWeaver.cs]: Maybe you could add a comment explaining the purpose of these two?

File: Weaver/RealmWeaver.Fody/ModuleWeaver.cs (hunk 2 above)
Comment (nirinchev) [inline Weaver/RealmWeaver.Fody/ModuleWeaver.cs]: I hadn't introduced the non-generic one at that point :) I think it's fine to leave it like that, but I'll need to ensure that the correct `MethodReference` is obtained.

**Comment by nirinchev** [comment]
retest this please

**Comment by nirinchev** [comment]
retest this please

**Pull request merged**